### PR TITLE
Fix 'zfs inherit snapdev' device node behaviour

### PR DESCRIPTION
**Pull request by loli10K**
### Description

When inheriting the 'snapdev' property to "visible" we don't call `zfs_prop_set_special()`: this prevents device nodes from being created.

Because 'snapdev' is the only "special" property that is also inheritable we need to call `zfs_prop_set_special()` even when we're not reverting to the received value (`zfs inherit -S`).

### Motivation and Context

I discovered this while porting FreeBSD "volmode" to ZoL: i was trying to test all the corner cases and found that inheriting snapdev was not triggering zvol_*_minor functions.

I'm kind of two minds about this: this is increasing the code difference between us and the other ZFS implementations, but on the other hand the 'snapdev' property is specific to ZoL so it's kind of inevitable to diverge until 'snapdev' is ported to other platforms.

```
root@debian-8-zfs:~# function block_device_wait() {
>    udevadm trigger
>    udevadm settle
> }
root@debian-8-zfs:~# #
root@debian-8-zfs:~# POOLNAME='testpool'
root@debian-8-zfs:~# TMPDIR='/var/tmp'
root@debian-8-zfs:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@debian-8-zfs:~# zpool destroy $POOLNAME
root@debian-8-zfs:~# rm -f $TMPDIR/zpool.dat
root@debian-8-zfs:~# fallocate -l 128m $TMPDIR/zpool.dat
root@debian-8-zfs:~# zpool create -O mountpoint=none $POOLNAME $TMPDIR/zpool.dat
root@debian-8-zfs:~# #
root@debian-8-zfs:~# zfs set snapdev=visible $POOLNAME
root@debian-8-zfs:~# zfs create -V 1M -o snapdev=hidden $POOLNAME/zvol1
root@debian-8-zfs:~# zfs snap $POOLNAME/zvol1@snap1
root@debian-8-zfs:~# block_device_wait
root@debian-8-zfs:~# ls -l /dev/zvol/$POOLNAME/
total 0
lrwxrwxrwx 1 root root 9 May 13 21:49 zvol1 -> ../../zd0
root@debian-8-zfs:~# # inherit 'snapdev'
root@debian-8-zfs:~# zfs inherit snapdev $POOLNAME/zvol1
root@debian-8-zfs:~# block_device_wait
root@debian-8-zfs:~# zfs get -o all snapdev $POOLNAME/zvol1
NAME            PROPERTY  VALUE    RECEIVED  SOURCE
testpool/zvol1  snapdev   visible  -         inherited from testpool
root@debian-8-zfs:~# # snapdev=visible, we should have 2 dev nodes
root@debian-8-zfs:~# ls -l /dev/zvol/$POOLNAME/
total 0
lrwxrwxrwx 1 root root 9 May 13 21:49 zvol1 -> ../../zd0
root@debian-8-zfs:~# # set snapdev=visible, dev node is created
root@debian-8-zfs:~# zfs set snapdev=visible $POOLNAME/zvol1
root@debian-8-zfs:~# zfs get -o all snapdev $POOLNAME/zvol1
NAME            PROPERTY  VALUE    RECEIVED  SOURCE
testpool/zvol1  snapdev   visible  -         local
root@debian-8-zfs:~# ls -l /dev/zvol/$POOLNAME/
total 0
lrwxrwxrwx 1 root root  9 May 13 21:49 zvol1 -> ../../zd0
lrwxrwxrwx 1 root root 10 May 13 21:49 zvol1@snap1 -> ../../zd16
root@debian-8-zfs:~# 
```

### How Has This Been Tested?

Test case added to zvol_misc

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: module/zfs/zfs_ioctl.c [hunk 1]
```diff
@@ -2816,8 +2774,49 @@ zfs_ioc_inherit_prop(zfs_cmd_t *zc)
...
+	pair = nvlist_next_nvpair(dummy, NULL);
+	if (pair == NULL) {
+		err = SET_ERROR(EINVAL);
+		goto errout;
+	} else {
+		err = zfs_prop_set_special(zc->zc_name, source, pair);
+	}
+
+	if (err == -1) /* property is not "special", needs to be handled */
+		err = dsl_prop_inherit(zc->zc_name, zc->zc_value, source);
```
Comment (behlendorf) [inline module/zfs/zfs_ioctl.c]: I'd be inclined to pull this block in to the `else` clause above.  I think that would make it clearer that `dsl_prop_inherit()` is called as the default behavior when no special behavior is defined by `zfs_prop_set_special()`.  But I'm OK with this either way.

**Comment by loli10K** [comment]
There's another small issue that only occurred to me while writing the test case: when we `zfs set snapdev` on the parent dataset the value is also *locally set* (not inherited) on the children:

```
root@debian-8-zfs:~# zfs create -V 1M $POOLNAME/zvol1
root@debian-8-zfs:~# #
root@debian-8-zfs:~# # snapdev=default => ok
root@debian-8-zfs:~# zfs get -o all snapdev -r $POOLNAME
NAME            PROPERTY  VALUE    RECEIVED  SOURCE
testpool        snapdev   hidden   -         default
testpool/zvol1  snapdev   hidden   -         default
root@debian-8-zfs:~# #
root@debian-8-zfs:~# # snapdev locally set on both => *not* ok
root@debian-8-zfs:~# zfs set snapdev=visible $POOLNAME
root@debian-8-zfs:~# zfs get -o all snapdev -r $POOLNAME
NAME            PROPERTY  VALUE    RECEIVED  SOURCE
testpool        snapdev   visible  -         local
testpool/zvol1  snapdev   visible  -         local
root@debian-8-zfs:~# #
root@debian-8-zfs:~# # inherit snapdev, verify it's inheritable => ok
root@debian-8-zfs:~# zfs inherit snapdev $POOLNAME/zvol1
root@debian-8-zfs:~# zfs get -o all snapdev -r $POOLNAME
NAME            PROPERTY  VALUE    RECEIVED  SOURCE
testpool        snapdev   visible  -         local
testpool/zvol1  snapdev   visible  -         inherited from testpool
root@debian-8-zfs:~# #
root@debian-8-zfs:~# # now set snapdev on parent again, value is also set locally on child zvol => *not* ok
root@debian-8-zfs:~# zfs set snapdev=visible $POOLNAME
root@debian-8-zfs:~# zfs get -o all snapdev -r $POOLNAME
NAME            PROPERTY  VALUE    RECEIVED  SOURCE
testpool        snapdev   visible  -         local
testpool/zvol1  snapdev   visible  -         local
```

I was hoping to push this without touching "zvol.c" but i don't see any way around it now since the issue is probably in zvol_set_snapdev* functions: this is going to take me a little more time to complete.

**Comment by loli10K** [comment]
@behlendorf the code to address the issue reported in my comment has been force-pushed and the test case updated accordingly: is there any other change that remains to be done?

**Comment by behlendorf** [comment]
@loli10K thanks.  Looks good, I'll go ahead and merge it.

**Comment by behlendorf** [comment]
@loli10K I'm going to revert this change from master.  Despite passing the testing in the PR it's causing a [NULL deref](http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20Release%20%28TEST%29/builds/2495/steps/shell_8/logs/log) after being merged to master we need to resolve.

```
01:36:24.96 [ 6636.713492] BUG: unable to handle kernel NULL pointer dereference at 0000000000000220
01:36:24.96 [ 6636.713552] IP: [<ffffffffa0422ead>] zvol_free+0xd/0x120 [zfs]
01:36:24.96 [ 6636.713554] PGD 1b0605067 
01:36:24.96 [ 6636.713554] PUD 11420a067 
01:36:24.96 [ 6636.713555] PMD 0 
01:36:24.96 
01:36:24.96 [ 6636.713557] Oops: 0000 [#1] SMP
01:36:24.96 [ 6636.713576] Modules linked in: sg(E) sd_mod(E) ext2(E) loop(E) zfs(POE) icp(POE) zcommon(POE) zunicode(POE) znvpair(POE) zavl(POE) splat(OE) spl(OE) nfsd(E) auth_rpcgss(E) nfs_acl(E) lockd(E) grace(E) sunrpc(E) ipv6(E) crc_ccitt(E) binfmt_misc(E) mousedev(E) evdev(E) psmouse(E) acpi_cpufreq(E) button(E) ext4(E) crc16(E) jbd2(E) fscrypto(E) mbcache(E) ata_piix(E) libata(E) scsi_mod(E) dm_mirror(E) dm_region_hash(E) dm_log(E) dm_mod(E) [last unloaded: scsi_debug]
01:36:24.96 [ 6636.713578] CPU: 0 PID: 3713 Comm: z_zvol Tainted: P           OE   4.9.27-14.31.amzn1.x86_64 #1
01:36:24.96 [ 6636.713579] Hardware name: Xen HVM domU, BIOS 4.2.amazon 11/11/2016
01:36:24.96 [ 6636.713580] task: ffff88010b50bb00 task.stack: ffffc90008910000
01:36:24.96 [ 6636.713625] RIP: 0010:[<ffffffffa0422ead>]  [<ffffffffa0422ead>] zvol_free+0xd/0x120 [zfs]
01:36:24.96 [ 6636.713626] RSP: 0018:ffffc90008913c68  EFLAGS: 00010246
01:36:24.96 [ 6636.713627] RAX: 0000000000000001 RBX: 0000000000000000 RCX: 0000000004448deb
01:36:24.96 [ 6636.713627] RDX: 0000000000000001 RSI: ffff88010ab56eaa RDI: 0000000000000000
01:36:24.96 [ 6636.713628] RBP: ffffc90008913c70 R08: 000000000001ade0 R09: ffffffffa0ba8d4a
01:36:24.96 [ 6636.713629] R10: ffff8801e7a1ade0 R11: ffffea00042ad580 R12: 0000000000000000
01:36:24.96 [ 6636.713630] R13: ffff88010ab56ea0 R14: fffffffffffffe40 R15: 00000000000001c0
01:36:24.96 [ 6636.713632] FS:  0000000000000000(0000) GS:ffff8801e7a00000(0000) knlGS:0000000000000000
01:36:24.96 [ 6636.713633] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
01:36:24.96 [ 6636.713634] CR2: 0000000000000220 CR3: 000000010fa22000 CR4: 00000000001406f0
01:36:24.96 [ 6636.713636] Stack:
01:36:24.96 [ 6636.713639]  ffff8801cd8a7800 ffffc90008913ca8 ffffffffa04248c2 ffff8800c0297000
01:36:24.96 [ 6636.713641]  ffff88010ab56ea0 ffffc90008913d00 ffff88011f418504 0000000000000001
01:36:24.96 [ 6636.713643]  ffffc90008913d60 ffffffffa037d6b0 000000000000006a 0000000000000000
01:36:24.96 [ 6636.713644] Call Trace:
01:36:24.96 [ 6636.713687]  [<ffffffffa04248c2>] zvol_set_snapdev_cb+0x152/0x170 [zfs]
01:36:24.96 [ 6636.713723]  [<ffffffffa037d6b0>] dmu_objset_find_impl+0x350/0x3d0 [zfs]
01:36:24.96 [ 6636.713765]  [<ffffffffa0424770>] ? zvol_task_cb+0x420/0x420 [zfs]
01:36:24.96 [ 6636.713805]  [<ffffffffa0424770>] ? zvol_task_cb+0x420/0x420 [zfs]
01:36:24.96 [ 6636.713840]  [<ffffffffa037d779>] dmu_objset_find+0x49/0x70 [zfs]
01:36:24.96 [ 6636.713881]  [<ffffffffa04243cd>] zvol_task_cb+0x7d/0x420 [zfs]
01:36:24.96 [ 6636.713885]  [<ffffffff8152ed54>] ? __schedule+0x244/0x680
01:36:24.96 [ 6636.713894]  [<ffffffffa0bac9fa>] taskq_thread+0x23a/0x470 [spl]
01:36:24.96 [ 6636.713897]  [<ffffffff8109b4e0>] ? wake_up_q+0x80/0x80
01:36:24.96 [ 6636.713903]  [<ffffffffa0bac7c0>] ? taskq_thread_spawn+0x50/0x50 [spl]
01:36:24.96 [ 6636.713906]  [<ffffffff8108fd67>] kthread+0xd7/0xf0
01:36:24.96 [ 6636.713908]  [<ffffffff8108fc90>] ? kthread_park+0x60/0x60
01:36:24.96 [ 6636.713911]  [<ffffffff815338c5>] ret_from_fork+0x25/0x30
01:36:24.96 [ 6636.713935] Code: 41 5e 41 5f 5d c3 89 c2 bb fb ff ff ff f7 da 83 f8 34 0f 45 da eb 8a e8 f2 82 c9 e0 eb aa 0f 1f 44 00 00 55 48 89 e5 53 48 89 fb <48> 8b 87 20 02 00 00 48 85 c0 0f 85 c9 00 00 00 48 8d bb 60 01 
01:36:24.96 [ 6636.713975] RIP  [<ffffffffa0422ead>] zvol_free+0xd/0x120 [zfs]
```

**Comment by loli10K** [comment]
@behlendorf i'm sorry for causing trouble, i'll try to troubleshoot and fix this as soon as possible 😢

**Comment by behlendorf** [comment]
@loli10K no problem.  I'm still not sure how the PR passed all the automated testing yet fails so reliably after being merged.  I'll be interested to hear what the root cause is.

**Comment by loli10K** [comment]
@behlendorf my local branch is missing both https://github.com/zfsonlinux/zfs/commit/07783588bcb513a3a1f4d995b5d4685a9cfc89e5 and https://github.com/zfsonlinux/zfs/commit/5559ba094feff560abe00afd31ab99dd1f70698c ZVOL related commits, i'll try to remember to rebase on master before the final push next time.

I've compiled my zfs.ko with debug info:
```
(gdb) bt
#0  0xffffffffc06a226f in zvol_free (arg=0x0 <irq_stack_union>) at /usr/src/zfs/module/zfs/zvol.c:1578
#1  0xffffffffc06a68fc in zvol_remove_minor_impl (name=0x0 <irq_stack_union>) at /usr/src/zfs/module/zfs/zvol.c:2049
#2  0xffffffffc06a6978 in zvol_set_snapdev_cb (dsname=0xffff88002b9ff480 "testpool/vol@snap", param=0xffff8802783d7dd8) at /usr/src/zfs/module/zfs/zvol.c:2112
#3  0xffffffffc05c3189 in dmu_objset_find_impl (spa=<optimized out>, name=0xffff8802928fed04 "testpool/vol", func=<optimized out>, arg=<optimized out>, flags=1) at /usr/src/zfs/module/zfs/dmu_objset.c:2397
#4  0xffffffffc05c3263 in dmu_objset_find (name=0xffff8802928fed04 "testpool/vol", func=0xffffffffc06a6910 <zvol_set_snapdev_cb>, arg=0xffff8802783d7dd8, flags=1) at /usr/src/zfs/module/zfs/dmu_objset.c:2431
#5  0xffffffffc06a4c9d in zvol_set_snapdev_impl (snapdev=<optimized out>, name=<optimized out>) at /usr/src/zfs/module/zfs/zvol.c:2128
#6  zvol_task_cb (param=0x0 <irq_stack_union>) at /usr/src/zfs/module/zfs/zvol.c:2181
#7  0xffffffffc02e95f3 in taskq_thread (args=0xffff8802861d8a80) at /usr/src/spl/module/spl/spl-taskq.c:921
#8  0xffffffff8109ac4f in kthread (_create=0xffff88028dae0380) at kernel/kthread.c:209
#9  0xffffffff815cc3f2 in ret_from_fork () at arch/x86/entry/entry_64.S:406
#10 0x0000000000000000 in ?? ()
(gdb) list zvol_free
1569	 * the zvol_state_list, and has its private data set to NULL.
1570	 * The zvol_state_lock is dropped.
1571	 */
1572	static void
1573	zvol_free(void *arg)
1574	{
1575		zvol_state_t *zv = arg;
1576	
1577		ASSERT(!MUTEX_HELD(&zvol_state_lock));
1578		ASSERT(zv->zv_open_count == 0);
(gdb) p zv
$13 = (zvol_state_t *) 0x0 <irq_stack_union>
(gdb) p zv->zv_open_count
Cannot access memory at address 0x11c
```

What seems strange to me is the change to `zvol_remove_minor_impl()` (from https://github.com/zfsonlinux/zfs/commit/5559ba094feff560abe00afd31ab99dd1f70698c):

```c
  2009  /* Remove minor for this specific snapshot only */                                                                                                                                               [353/1810]
  2010  static void
  2011  zvol_remove_minor_impl(const char *name)
  2012  {
  2013          zvol_state_t *zv, *zv_next;
  2014
  2015          if (zvol_inhibit_dev)
  2016                  return;
  2017
  2018          if (strchr(name, '@') == NULL)
  2019                  return;
  2020
  2021          mutex_enter(&zvol_state_lock);
  2022
  2023          for (zv = list_head(&zvol_state_list); zv != NULL; zv = zv_next) {
  2024                  zv_next = list_next(&zvol_state_list, zv);
  2025
  2026                  if (strcmp(zv->zv_name, name) == 0) {
  2027                          /* If in use, leave alone */
  2028                          if (zv->zv_open_count > 0 ||
  2029                              atomic_read(&zv->zv_suspend_ref))
  2030                                  continue;
  2031                          /*
  2032                           * By taking zv_state_lock here, we guarantee that no
  2033                           * one is currently using this zv
  2034                           */
  2035                          mutex_enter(&zv->zv_state_lock);
  2036                          zvol_remove(zv);
  2037                          mutex_exit(&zv->zv_state_lock);
  2038                          /* clear this so zvol_open won't open it */
  2039                          zv->zv_disk->private_data = NULL;
  2040                          break;
  2041                  }
  2042          }
  2043
  2044          mutex_exit(&zvol_state_lock);
  2045
  2046          /*
  2047           * Drop zvol_state_lock before calling zvol_free()
  2048           */
  2049          zvol_free(zv);
```

What happens when `zvol_state_list` is empty? Isn't `zv` still uninitiated when do the `zvol_free(zv)`?

Before https://github.com/zfsonlinux/zfs/commit/5559ba094feff560abe00afd31ab99dd1f70698c we did `zvol_free(zv)` inside the for loop, which seems safer to me.

**Comment by behlendorf** [comment]
> What happens when zvol_state_list is empty?

That could definitely cause this bug and we need to add check for NULL.  It appears we didn't noticed since without your change we didn't have test coverage for this.  Let's get @bprotopopov's thoughts on this too.

**Comment by bprotopopov** [comment]
@behlendorf, why would zvol_remove_minor_impl() be called if there are no minors to remove (zvol_state_list is empty) ? Is this possible ?

**Comment by bprotopopov** [comment]
@loli10K, @behlendorf 
In the call stack above:
```
...
#0  0xffffffffc06a226f in zvol_free (arg=0x0 <irq_stack_union>) at /usr/src/zfs/module/zfs/zvol.c:1578
#1  0xffffffffc06a68fc in zvol_remove_minor_impl (name=0x0 <irq_stack_union>) at /usr/src/zfs/module/zfs/zvol.c:2049
...
```
how can the name for #1 be NULL ? Looking at the code in `zvol_snapdev_change_cb() `, I can't see how this is possible.

Regarding check for zv == NULL, I think it's a good idea as it makes the code easy to understand. I'll look more to see if this is strictly needed.

**Comment by bprotopopov** [comment]
@behlendorf, @loli10K, yes, it seems like it's a good idea to check for NULL. The `zvol_remove_minor_impl()` is called directly from the callback that changes snapdev, so, if it's called more than once setting snapdev to 'hidden', and then zv will be NULL;  zvol_state_list does not need to be empty.

**Comment by loli10K** [comment]
> The zvol_remove_minor_impl() is called directly from the callback that changes snapdev, so, if it's called more than once setting snapdev to 'hidden', and then zv will be NULL

Confirmed, i can reproduce the same failure on https://github.com/zfsonlinux/zfs/commit/5559ba094feff560abe00afd31ab99dd1f70698c settings `snapdev=hidden` twice:

```
root@debian-8-zfs:~# cat /sys/module/zfs/version 
0.7.0-rc4_14_g5559ba094
root@debian-8-zfs:~# POOLNAME='testpool'
root@debian-8-zfs:~# TMPDIR='/var/tmp'
root@debian-8-zfs:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@debian-8-zfs:~# zpool destroy $POOLNAME
cannot open 'testpool': no such pool
root@debian-8-zfs:~# rm -f $TMPDIR/zpool.dat
root@debian-8-zfs:~# fallocate -l 128m $TMPDIR/zpool.dat
root@debian-8-zfs:~# zpool create -O mountpoint=none $POOLNAME $TMPDIR/zpool.dat
root@debian-8-zfs:~# zfs create -V 1M -o snapdev=hidden $POOLNAME/zvol1
root@debian-8-zfs:~# zfs snap $POOLNAME/zvol1@snap1
root@debian-8-zfs:~# zfs set snapdev=hidden $POOLNAME
[  182.876102] BUG: unable to handle kernel NULL pointer dereference at 000000000000011c
[  182.879639] IP: [<ffffffffc069a042>] zvol_free+0x22/0x1e0 [zfs]
[  182.882660] PGD 0 
[  182.883152] Oops: 0000 [#1] SMP 
[  182.883919] Dumping ftrace buffer:
[  182.884701]    (ftrace buffer empty)
```

> So, perhaps the tests that set snapdev should do this - set snapsev to the same value (visible/hidden) more than once in a row.

I think my commit did exactly this, that's why it was triggering this bug.

**Comment by bprotopopov** [comment]
OK, I can commit this fix separately if needed or @loli10K can include it with the rest of the code changes ?

**Comment by bprotopopov** [comment]
@behlendorf , I am pretty sure there were whole bunch of tests in zfs-tests that set snapdev to visible and hidden and then checked for the device links. I can't seem t find any at this point.

**Comment by bprotopopov** [comment]
@loli10K @behlendorf, I would prefer to do the following:

- adding check for the zv == NULL
- making sure we are not doing anything if snapdev is set to the value it currently holds (if @loli10K commit will make this commonplace, it seems it makes sense to optimize here)
- adding some tests that set snapdev property

**Comment by bprotopopov** [comment]
I think I found a good test to extend: 
`tests/zfs-tests/tests/functional/zvol/zvol_misc/zvol_misc_snapdev.ksh`

**Comment by loli10K** [comment]
> making sure we are not doing anything if snapdev is set to the value it currently holds

This seems a little more fragile than the current implementation.

For instance setting `volsize=1M` multiple times actually "update" the zap entry "size" (and "refreservation") with the same value:

```
root@debian-8-zfs:~# POOLNAME='testpool'
root@debian-8-zfs:~# TMPDIR='/var/tmp'
root@debian-8-zfs:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@debian-8-zfs:~# zpool destroy $POOLNAME
root@debian-8-zfs:~# rm -f $TMPDIR/zpool.dat
root@debian-8-zfs:~# fallocate -l 128m $TMPDIR/zpool.dat
root@debian-8-zfs:~# zpool create -O mountpoint=none $POOLNAME $TMPDIR/zpool.dat
root@debian-8-zfs:~# #
root@debian-8-zfs:~# zfs create -V 1M $POOLNAME/zvol1
root@debian-8-zfs:~# #
root@debian-8-zfs:~# stap -d zfs -d spl --ldd \
>    -e '
> probe
> module("zfs").function("zvol_update_volsize").call
> {
>    printf("-> %s (volsize=%s)\n", ppfunc(), $volsize$);
> }
> probe
> module("zfs").function("zap_update").call
> {
>    printf("-> %s (name=%s, value=%s)\n", ppfunc(), $name$, $val$);
> }' -c bash -c "
> for i in 1 2 3 4 5; do zfs set volsize=1M $POOLNAME/zvol1; done"
-> zvol_update_volsize (volsize=1048576)
-> zap_update (name="size", value=0xffff8800cba3fca8)
-> zap_update (name="refreservation", value=0xffff8801fe4dfbf0)
-> zvol_update_volsize (volsize=1048576)
-> zap_update (name="size", value=0xffff8800cba07ca8)
-> zap_update (name="refreservation", value=0xffff8801fe4dfbf0)
-> zvol_update_volsize (volsize=1048576)
-> zap_update (name="size", value=0xffff88020ca57ca8)
-> zap_update (name="refreservation", value=0xffff8801fe4dfbf0)
-> zvol_update_volsize (volsize=1048576)
-> zap_update (name="size", value=0xffff8801ffa07ca8)
-> zap_update (name="refreservation", value=0xffff8801fe4dfbf0)
-> zvol_update_volsize (volsize=1048576)
-> zap_update (name="size", value=0xffff88020c863ca8)
-> zap_update (name="refreservation", value=0xffff8801fe4dfbf0)
```
I think we should deal with `snapdev` the same way we do for all other ZVOL properties for sake of consistency, so allow the same value to be set multiple times. That said, the ZVOL code is still very new to me, so i'm fine with whatever option you choose.

**Comment by bprotopopov** [comment]
@loli10K, unlike other properties, setting `snapdev` causes scan for all snapshots of the zvol. That will probably not scale well with large numbers of zvols an snapshots.

**Comment by loli10K** [comment]
> That will probably not scale well with large numbers of zvols and snapshots.

That is true, but if we decide our actions based on what _we think_ the situation is (read: what's the property value) we could find ourselves in strange situations. Consider the following case:

1. echo 1 > /sys/module/zfs/parameters/zvol_inhibit_dev
2. zfs set snapdev=visible testpool/zvol (now _we think_ our zvol is "visible", it's actually hidden by `zvol_inhibit_dev`)
3. echo 0 > /sys/module/zfs/parameters/zvol_inhibit_dev
4. zfs set snapdev=visible testpool/zvol

If we "make sure we are not doing anything if snapdev is set to the value it currently holds" step 4 will not create any device node, which is not what i would expect.

**Comment by bprotopopov** [comment]
@loli10K 

I don't think `zvol_inhibit_dev` is designed to be used in this fashion. This is more of a one-time setting for a given storage stack: either make zvols visible as block devices or not. This is a system-design-time decision, not a runtime decision, to my knowledge.

I believe that the scenario you have outlined above is unlikely to be important in practice. And if it is important is some rare cases, there is a simple workaround:
1. echo 1 > /sys/module/zfs/parameters/zvol_inhibit_dev
2. zfs set snapdev=visible testpool/zvol (now we think our zvol is "visible", it's actually hidden by zvol_inhibit_dev)
3. echo 0 > /sys/module/zfs/parameters/zvol_inhibit_dev
4. zfs set snapdev=hidden testpool/zvol
5. zfs set snapdev=visible testpool/vol

Rather than optimizing for these rare and non-intuitive cases, I would prefer to optimize for the case that arises as a result of your changes. In other words, I would like to understand better how and why the `snappev` property is being set to hidden multiple times. Can you please elaborate ?

**Comment by loli10K** [comment]
> In other words, I would like to understand better how and why the snappev property is being set to hidden multiple times. Can you please elaborate ?

We have a ZFS backup host with one backup pool: this storage is used to receive virtual machine backups, which are deployed on ZVOLs on the primary storage.

Every time we receive a new volume another device node is created, and since we don't want /dev cluttered with zd devices, we set `zvol_inhibit_dev=1`.

Everything works fine until one day we get a request to restore a file from one volume snapshot, so we promptly set `snapdev=visible` and go to access the `/dev/zvol/backuppool/vol@snap` device, which we don't find because we set `zvol_inhibit_dev=1`.
Ok, no worries, we set `zvol_inhibit_dev=0` and try to access the "snapshot" device node again, which we still don't find.
Now we remember reading somewhere that `zvol_inhibit_dev` is not a one-time setting: do we need to export/import the pool? Do we need to reboot the backup host with `zvol_inhibit_dev=0`? Maybe the request to restore the backup was not urgent, maybe it was and the phone starts ringing. Either way, we try to set `snapdev=visible` again on the volume as the last resort and it still doesn't work (for the reasons we discussed).
Yes, there is a simple workaround as you said, which you hopefully come up with (or remember reading) when you need it. Chances are your head is not cool enough in that situation and you end up rebooting the node, which will hopefully it come back up fast (if we really need that snapshot device), even if now we have to create *all* the zvol device nodes on pool import.

**Comment by bprotopopov** [comment]
I was actually asking about your fix to the inherit feature causing multiple sets of snapdev. How does that happen?

Typos courtesy of my iPhone

On May 27, 2017, at 1:37 AM, LOLi <notifications@github.com<mailto:notifications@github.com>> wrote:


In other words, I would like to understand better how and why the snappev property is being set to hidden multiple times. Can you please elaborate ?

We have a ZFS backup host with one backup pool: this storage is used to receive virtual machine backups, which are deployed on ZVOLs on the primary storage.

Every time we receive a new volume another device node is created, and since we don't want /dev cluttered with zd devices, we set zvol_inhibit_dev=1.

Everything works fine until one day we get a request to restore a file from one volume snapshot, so we promptly set snapdev=visible and go to access the /dev/zvol/backuppool/vol@snap device, which we don't find because we set zvol_inhibit_dev=1.
Ok, no worries, we set zvol_inhibit_dev=0 and try to access the "snapshot" device node again, which we still don't find.
Now we remember reading somewhere that zvol_inhibit_dev is not a one-time setting: do we need to export/import the pool? Do we need to reboot the backup host with zvol_inhibit_dev=0? Maybe the request to restore the backup was not urgent, maybe it was and the phone starts ringing. Either way, we try to set snapdev=visible again on the volume as the last resort and it still doesn't work (for the reasons we discussed).
Yes, there is a simple workaround as you said, which you hopefully come up with (or remember reading) when you need it. Chances are your head is not cool enough in that situation and you end up rebooting the node, which will hopefully it come back up fast (if we really need that snapshot device), even if now we have to create all the zvol device nodes on pool import.

**Comment by bprotopopov** [comment]
@loli10K 
if the above scenario is not hypothetical :) 
my suggestion is to set zvol_inhibit_dev=0, snapdev=hidden on all your zvols (these are defaults). This will only cause device nodes for zvols to be created, but not for the snapshots, and this scales to thousands of zvols, to my knowledge.
Then, when you need to temporarily access your snapshot as device for restore, simply clone that snapshot (set read-only on the device if you need to), and the device node for the clone will be created. Once you are done with the restore, destroy the clone, and the device node will go away.

P.S. To clarify, I am trying to understand the multiple setting of the snapdev property in the context of your code changes.

**Comment by bprotopopov** [comment]
@loli10K , @behlendorf 
I don't have strong feelings as to checking for the current value of the snapdev before setting it and triggering the minor functions, but it would be nice to tighten this a bit.

**Comment by bprotopopov** [comment]
@loli10K 

from your comments above:
```
...
root@debian-8-zfs:~# # snapdev locally set on both => *not* ok
root@debian-8-zfs:~# zfs set snapdev=visible $POOLNAME
root@debian-8-zfs:~# zfs get -o all snapdev -r $POOLNAME
NAME            PROPERTY  VALUE    RECEIVED  SOURCE
testpool        snapdev   visible  -         local
testpool/zvol1  snapdev   visible  -         local
...
```
why is this not OK ? I thought that snapdev was a property that was inherited automagically.

**Comment by loli10K** [comment]
> inherited

Exactly, the SOURCE column for "_testpool/zvol1_" should report "_inherited from testpool_" if is was inherited; it is (improperly) set as a "local" value instead, which is not ok.
